src/data_readers/data_reader_hdf5.cpp

```diff
@@ -0,0 +1,237 @@
+                offset[1] = ((yPerNode+1)*(rank%num_io_parts));
+             } else {
+                // offset of the y dim for this rank
+                odd_offset = (yPerNode+1)*(dims[1]%yPerNode);
+                offset[1] = odd_offset + (yPerNode*((rank%num_io_parts)/ylines));
+             }   
+        } else {
+            offset[1] = 0;
+        }
+
+        if (zlines > 1) {
+            offset[2] = zPerNode*(rank%num_io_parts);
+        } else {
+            offset[2] = 0;
+        }
+        // I have channel dim splits 
+        // all combos arent really tested
+        //add the rest later
+        offset[3] = 0; 
+        count[0] = 1;
+        count[1] = 1;
+        count[2] = 1;
+        count[3] = 1;
+        // from an explanation of the hdf5 select_hyperslab:
+        // start -> a starting location for the hyperslab
+        // stride -> the number of elements to separate each element or block to be selected
+        // count -> the number of elemenets or blocks to select along each dimension
+        // block -> the size of the block selected from the dataspace 
+        //hsize_t status;
+        
+        //todo add error checking
+        H5Sselect_hyperslab(filespace, H5S_SELECT_SET, offset, NULL, count, dims_local);
+        H5Dread(h_data, H5T_NATIVE_SHORT, memspace, filespace, H5P_DEFAULT, data_out);
+    }
+
+    void hdf5_reader::load() {
+       
+        std::string dirpath = get_file_dir();    
+        std::vector<std::string> file_list = get_filenames(dirpath);
+        lbann_comm* l_comm = get_comm();
+        const El::mpi::Comm & w_comm = l_comm->get_world_comm();
+        MPI_Comm mpi_comm = w_comm.GetMPIComm();
+        int world_rank = get_rank_in_world();
+        int color = world_rank/dc::get_number_of_io_paritions(); 
+        MPI_Comm_split(mpi_comm, color, world_rank, &m_comm);
+        m_shuffled_indices.clear();
+        m_shuffled_indices.resize(m_file_paths.size());
+        std::iota(m_shuffled_indices.begin(), m_shuffled_indices.end(), 0);
+        std::cout<<"size after load " << m_shuffled_indices.size();
+        
+        select_subset_of_data();
+    }
+    bool hdf5_reader::fetch_label(Mat& Y, int data_id, int mb_idx) {
+        return true;
+    }
+    bool hdf5_reader::fetch_datum(Mat& X, int data_id, int mb_idx) {
+        prof_region_begin("fetch_datum", prof_colors[0], false);
+        int nprocs;
+        MPI_Comm_size(MPI_COMM_WORLD, &nprocs); 
+        if ((nprocs%dc::get_number_of_io_partitions()) !=0) {
+            std::cerr<<"if other things have not been changed in the code this will not work for anything other than 4 procs a node \n";
+        }
+        int world_rank = get_rank_in_world();
+        double start_file = MPI_Wtime();
+        //TODO: do I need this mod --> will world rank/paritions ever be greater than the number of files??
+	    auto file = m_file_list[((world_rank/dc::get_number_of_io_partitions())+nux)%(m_file_list.size())];
```

Comment:
`data_id` is the index of the sample to read. no need to calculate here. you can assume it's the same with a group of `dc::get_number_of_io_partitions()` procs.